src/enumerable.cr

```diff
@@ -636,11 +636,7 @@ module Enumerable(T)
     h = Hash(U, Array(T)).new
     each do |e|
       v = yield e
-      if h.has_key?(v)
-        h[v].push(e)
-      else
-        h[v] = [e]
-      end
+      h.put_if_absent(v, Array(T).new) << e
```

Comment:
This unconditionally constructs a new array on every iteration. Instead do:

```suggestion
      h.put_if_absent(v) { Array(T).new } << e
```

or split it into 2 expressions